### PR TITLE
Upgrade build_vendors_travis.sh to be usable also for local vendors preparation in unix base system.

### PR DESCRIPTION
**Pull request by velmafia**
I don't like run manually what can be run automatically, so I upgrade build_vendors_travis.sh for unix base system, here is my proposition (pre-review).

My questions:
1. Why travis-ci target don't use libthmbnlr and translations (_src/xpiks-qt/deps/translations/_)?
2. Where it should be default libs directory, `libs`, or `libs/debug` and `libs/release`.

aloha!

### CODE REVIEW

File: vendors/build_vendors_unix.sh [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        TARGET="debug"
+        ;;
+    *)
+        echo "$(basename ${0}) error: Invalid command argument"
+        echo "    usage: $(basename ${0}) <debug|release|travis-ci>"
+        exit 1
+        ;;
+esac
+
+MAKE_FLAGS="-j$(nproc)"
```
Comment (velmafia) [inline vendors/build_vendors_unix.sh]: MAKE_FLAGS will be set by script arguments in next iteration.

**Comment by ribtoks** [comment]
Hi

1. _travis-ci_ just does not need libthmbnlr and translations to run tests at all. Those are only UI features needed for customers.
2. Default libs directory for debug is `libs/debug` and for release is `libs/release`. There's no other directory than "default"

File: .travis.yml [hunk 2]
```diff
@@ -60,7 +60,7 @@ before_script:
   - chmod +x src/xpiks-tests/get_code_cov.sh
 
 script:
-  - if [ -z "${UI_TESTS+xxx}" ]; then bash vendors/build_vendors_travis.sh; fi
+  - if [ -z "${UI_TESTS+xxx}" ]; then bash vendors/build_vendors_travis.sh travis-ci; fi
```
Comment (ribtoks) [inline .travis.yml]: Why is your script called `build_vendors_unix` and here you still execute `build_vendors_travis`?

File: vendors/build_vendors_unix.sh [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+ROOT_DIR=$(git rev-parse --show-toplevel)
+cd ${ROOT_DIR}
+
+if [ ${1} != "travis-ci" ]
```
Comment (ribtoks) [inline vendors/build_vendors_unix.sh]: It's better to put `${1}` to separate variable with descriptive name for better readability. Something like `$BUILD_MODE`

File: vendors/build_vendors_unix.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+then
+    ### submodule
+    git submodule init && git submodule update --recursive
+
+    ### translactions
```
Comment (ribtoks) [inline vendors/build_vendors_unix.sh]: Typo: translations

File: vendors/build_vendors_unix.sh [hunk 5]
```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
```
Comment (ribtoks) [inline vendors/build_vendors_unix.sh]: This script is useless on OS X so don't call it _build - unix . sh_. Let's better use _linux_. `build_vendors_linux.sh`

File: vendors/build_vendors_unix.sh [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+### hunspell
+cd ${ROOT_DIR}/vendors/hunspell/
+qmake "CONFIG+=${TARGET}" hunspell.pro
+make ${MAKE_FLAGS}
+
+### libxpks
+cd ${ROOT_DIR}/src/libxpks_stub
+qmake "CONFIG+=${TARGET}" libxpks_stub.pro
+make ${MAKE_FLAGS}
+cp libxpks.a ${ROOT_DIR}/libs/${TARGET}
```
Comment (ribtoks) [inline vendors/build_vendors_unix.sh]: All shell scripts should proudly `echo 'Done'` in the end :)

File: .travis.yml (hunk 2 above)
Comment (velmafia) [reply .travis.yml]: Fixed, yesterday was to late...
Why travis build was green - I also want to know :)

File: vendors/build_vendors_unix.sh (hunk 5 above)
Comment (velmafia) [reply vendors/build_vendors_unix.sh]: Changed.

File: vendors/build_vendors_unix.sh (hunk 3 above)
Comment (velmafia) [reply vendors/build_vendors_unix.sh]: changed

File: vendors/build_vendors_unix.sh (hunk 4 above)
Comment (velmafia) [reply vendors/build_vendors_unix.sh]: corrected

File: vendors/build_vendors_unix.sh (hunk 6 above)
Comment (velmafia) [reply vendors/build_vendors_unix.sh]: Done :)

**Comment by velmafia** [comment]
Hi,

1. I ask about this, because I prefer to build everything after commit to find potential problem immediately. For example, if I break `libthmbnlr` compilation, you don't find this in travis. 
Also my question is related to my first xpiks compilation, build instruction no mention about translation, and I couldn't find `libthmbnlr` project. My first thought is "where I making mistake that code is not compiling",  I spend some time to understand where I make mistake...
2. This questions is related to actual travis fail with my pull request (`LD_LIBRARY_PATH` is set to `libs/` directory, not to `libs/debug` or `libs/release`). Actual script (`build_vendors_travis.sh`) is copy all libraries to `libs/` (not `libs/debug` and `libs/release`) directory, also .pro files have set `LIBS += -L"$$PWD/../../../libs"` in some places (for example in `xpiks-tests-integration.pro` for `linux`, `travis-ci`, and `appveyor`).

File: src/xpiks-tests/xpiks-tests-integration/xpiks-tests-integration.pro [hunk 7]
```diff
@@ -590,7 +590,7 @@ win32 {
 
 linux {
     message("for Linux")
-    INCLUDEPATH += "../../vendors/quazip"
+    INCLUDEPATH += "../../../vendors/quazip"
     LIBS += -L"$$PWD/../../libs"
```
Comment (ribtoks) [inline src/xpiks-tests/xpiks-tests-integration/xpiks-tests-integration.pro]: So I didn't quite understand are you changing it to be consistent **everywhere** to point to `libs/ DEBUG|RELEASE` or not?

File: .travis.yml [hunk 8]
```diff
@@ -56,13 +56,13 @@ before_script:
...
-  - chmod +x vendors/build_vendors_travis.sh
+  - chmod +x vendors/build_vendors_linux.sh
   - chmod +x src/xpiks-tests/get_code_cov.sh
 
 script:
-  - if [ -z "${UI_TESTS+xxx}" ]; then bash vendors/build_vendors_travis.sh; fi
+  - if [ -z "${UI_TESTS+xxx}" ]; then bash vendors/build_vendors_linux.sh travis-ci; fi
   - cd src/
-  - ls -la ../libs/
+  - ls -la ../libs/*
```
Comment (ribtoks) [inline .travis.yml]: If files are in the `libs/debug` this command will be useless

**Comment by ribtoks** [comment]
@velmafia 

You don't need to rebuild vendors after each commit. Vendors change once or twice per year when I do the release and just update all 3rdparties to their latest stable released versions. Therefore this automation is useful of course, but it will not save much of your time everyday.

Again, you don't need translations for build. Building translations is part of _release_ procedures, not build procedures. Nothing in tests or elsewhere _requires_ translations to be present.

You couldn't find `libthmbnlr` because it's simply not in the repo, not required for the build, not required for the tests. That's why there's `libthmbnlr_stub.cpp` file - just to satisfy header declarations.

Regarding _libs_. In the ancient times everything just lived in the `libs/` directory, there were no separation to _debug_ and _release_. And travis build script used this approach (so you can see references to old way inside projects and `LD_LIBRARY_PATH`. Then I thought it's nice to separate those libs and locally did this separation. 

I didn't touch `travis` buildscript just because it worked well (and I don't like changing things which work well and there is literally **no** real necessity to change them). Just look at the old script and your "improved" one side by side. Which one looks easier to understand and maintain?

**Comment by ribtoks** [comment]
@velmafia Travis build succeeded because `vendors/` directory is cached. So currently we couldn't check if the script actually works. I've cleared the caches so hope to see if it works.

File: .travis.yml (hunk 8 above)
Comment (velmafia) [reply .travis.yml]: In my system (Debian) this command print also subdirectory, in travis the same:

> 0.01s$ ls -la ../libs/*
> ../libs/debug:
> total 17556
> drwxrwxr-x 2 travis travis    4096 Jan 20 08:19 .
> drwxrwxr-x 4 travis travis    4096 Jan 20 08:13 ..
> -rw-rw-r-- 1 travis travis       0 Jan 20 08:13 .keep
> -rw-rw-r-- 1 travis travis 1009400 Jan 20 08:18 libface.a
> -rw-rw-r-- 1 travis travis 4807326 Jan 20 08:19 libhunspell.a
> lrwxrwxrwx 1 travis travis      18 Jan 20 08:19 libquazip.so -> libquazip.so.1.0.0
> lrwxrwxrwx 1 travis travis      18 Jan 20 08:19 libquazip.so.1 -> libquazip.so.1.0.0
> lrwxrwxrwx 1 travis travis      18 Jan 20 08:19 libquazip.so.1.0 -> libquazip.so.1.0.0
> -rwxrwxr-x 1 travis travis 2309877 Jan 20 08:19 libquazip.so.1.0.0
> lrwxrwxrwx 1 travis travis      17 Jan 20 08:18 libssdll.so -> libssdll.so.1.0.0
> lrwxrwxrwx 1 travis travis      17 Jan 20 08:18 libssdll.so.1 -> libssdll.so.1.0.0
> lrwxrwxrwx 1 travis travis      17 Jan 20 08:18 libssdll.so.1.0 -> libssdll.so.1.0.0
> -rwxrwxr-x 1 travis travis  923953 Jan 20 08:18 libssdll.so.1.0.0
> -rw-rw-r-- 1 travis travis 8909002 Jan 20 08:19 libxpks.a
> ../libs/release:
> total 8
> drwxrwxr-x 2 travis travis 4096 Jan 20 08:13 .
> drwxrwxr-x 4 travis travis 4096 Jan 20 08:13 ..
> -rw-rw-r-- 1 travis travis    0 Jan 20 08:13 .keep

so it's make what I want.

File: src/xpiks-tests/xpiks-tests-integration/xpiks-tests-integration.pro (hunk 7 above)
Comment (velmafia) [reply src/xpiks-tests/xpiks-tests-integration/xpiks-tests-integration.pro]: ad. INCLUDEPATH:
I change this, to fix compilation
> ../../xpiks-qt/Helpers/ziphelper.cpp:13:10: fatal error: quazip/JlCompress.h: No such file or directory

for another targets path is correct, for linux is not. The same bug is with LIBS.

ad. libs:
I like to be consistent, and I can change this everywhere in project to `libs/ DEBUG|RELEASE`
(if it's not a problem for you).

**Comment by velmafia** [comment]
Ok, maybe I write my first attempt with xpiks build:

I'm a new here and I want contribute xpiks. I use Linux system. 
- I find build instruction ( https://github.com/ribtoks/xpiks/wiki/Build-instructions ) and I'm using it.
- I install system dependences: Qt, git - OK
- Check out Xpiks - OK
- Check out submodules - OK
- rename `src/tiny-AES/aes.c` - OK
- build dependencies (`Hunspell`, `Quazip`, `cpp-libface` ,and another from instruction) - OK

Now I try build main application:
- open `src/xpiks-qt/xpiks-qt.pro` - OK
- select Debug, and build: 
> cp: cannot stat '/home/vel/code/xpiks/src/xpiks-common/../xpiks-qt/deps/translations/xpiks_*.qm': No such file or directory
> Makefile:1434: recipe for target 'copytranslations' failed
- find where is problem, fix it (simply make in translations) - no info in build instruction about this
- build again:
> /usr/bin/ld: cannot find -lthmbnlr
> /usr/bin/ld: cannot find -lavcodec
> /usr/bin/ld: cannot find -lavfilter
> /usr/bin/ld: cannot find -lavformat
> (...)
- find where is problem, create `thmbnlr` project, install necessary libs in system - no info in build instruction about this
- build again - OK


Also I have problem with libs in `xpiks-tests-integration`, libs available in `libs/debug` and `lib/release` is available for linker, in `libs/` - is not (in linux system). 
Yes - this is bug, and I can fix it, but it's good opportunity to unify libs directory to whole project, 

Finally, now I can build, run, and work with xpiks repo, but I want simplify this procedure for new contributors. `build_vendors_linux.sh` is good place for this, because running by travis is guarantee that this script will be actually.

And I know that previous script was easier to understand, but without echo and "argument part", this is almost the same, and it's usable not only for travis.

---
> So currently we couldn't check if the script actually works. I've cleared the caches so hope to see if it works.

I tested it in my travis (with cleared cache) - it's working: https://travis-ci.org/velmafia/xpiks/builds/331071986

**Comment by ribtoks** [comment]
@velmafia Thanks for the explanations - all points pretty valid, I have nothing to add.

File: src/xpiks-tests/xpiks-tests-integration/xpiks-tests-integration.pro (hunk 7 above)
Comment (ribtoks) [reply src/xpiks-tests/xpiks-tests-integration/xpiks-tests-integration.pro]: @velmafia yes please, consistency is always good

**Comment by ribtoks** [comment]
@velmafia btw regarding translations: another _correct_ way to fix this build impediment is to fix the build step which copies .qm files to check if they are there (e.g. `cp *.qm || echo "no translations"`)

**Comment by velmafia** [comment]
> @velmafia btw regarding translations: another correct way to fix this build impediment is to fix the build step which copies .qm files to check if they are there (e.g. cp *.qm || echo "no translations")

OK, I will fix this together with `libs/ DEBUG|RELEASE` in a few days.

**Pull request merged**